beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -31,36 +37,281 @@ Introducing the rollback to scaffolder actions provides the mean to come back to
+
+### Securing Rollbacks
+
+We will provide a helper function that encapsulates the logic of verifying the resource and executing the rollback. Core security checks can be performed in this function to ensure the resource is
+valid and can be rolled back. There is a possible opportunity to allow others to extend this function to provide custom security checks.
+
+By default, we can handle the following:
+
+1. Verify we reached the point of setting output
+1. Parse resource details from the output and validate them against the input
+1. Execute the actual rollback with the validated resource
+
+```typescript
+export function createSecureRollback
+  TActionInput extends JsonObject,
+  TActionOutput extends JsonObject,
+  TSchemaType extends 'v1' | 'v2'
+>(
+  rollbackFn: (
+    ctx: ActionContext<TActionInput, TActionOutput, TSchemaType>,
+    resource: { id: string; type: string; metadata?: JsonObject }
+  ) => Promise<void>,
+  options: {
+    // Required resource identifier from the output
+    outputKey: keyof TActionOutput;
+    // Function to parse resource details from the output
+    parseResource: (value: JsonValue) => { id: string; type: string; metadata?: JsonObject };
+    // Validation to ensure resource matches input
+    validateResource: (
+      resource: { id: string; type: string; metadata?: JsonObject },
+      input: TActionInput
+    ) => boolean;
+  }
+) {
+  return async (ctx: ActionContext<TActionInput, TActionOutput, TSchemaType>) => {
+    // 1. Verify we reached the point of setting output
+    const outputValue = ctx.output?.values?.[options.outputKey];
+    if (outputValue === undefined) {
+      ctx.logger.info(`Skipping rollback: no resource was created (no value for ${String(options.outputKey)} in output)`);
+      return;
+    }
+
+    // 2. Parse resource details from the output
+    const resource = options.parseResource(outputValue);
+
+    // 2. Verify the resource matches what would have been created from the input
+    if (!options.validateResource(resource, ctx.input)) {
+      ctx.logger.error(
+        `Security check failed: output resource ${resource.id} doesn't match expected value from input parameters`
+      );
+      throw new Error(`Resource validation failed during rollback of ${resource.type}`);
+    }
+
+    try {
+      // 4. Execute the actual rollback with the validated resource
+      ctx.logger.info(`Rolling back ${resource.type} with id ${resource.id}`);
+      await action.rollback(ctx);
+      ctx.logger.info(`Successfully rolled back ${resource.type} with id ${resource.id}`);
+    } catch (error) {
+      ctx.logger.error(
+        `Failed to rollback ${resource.type} with id ${resource.id}: ${error}`,
+      );
+      ctx.logger.info('Continuing with next rollback...');
```

Comment:
This part might be configurable. Maybe by using a different strategy on a task level.

Reply:
@acierto - are you referring to the `TaskRecoverStrategy` [type](https://github.com/backstage/backstage/blob/403bb43d3dbcbe35ddc23a6530eddb7ceabcff9f/plugins/scaffolder-common/src/TaskSpec.ts#L54)?

Reply:
I'm now leaning towards this being a separate mechanism. Recovery and rollbacks are related, but they are distinct concerns.

Reply:
I was thinking more of creating a new one `TaskRollbackStrategy`:`StopOnFailure`, `ContinueOnFailure`, more cases?.